enterprise/server/remote_execution/snaploader/snaploader.go

```diff
@@ -182,24 +208,45 @@ func New(env environment.Env) (*FileCacheLoader, error) {
 	return &FileCacheLoader{env: env}, nil
 }
 
-func (l *FileCacheLoader) GetSnapshot(ctx context.Context, key *fcpb.SnapshotKey) (*Snapshot, error) {
-	var manifest *fcpb.SnapshotManifest
-	var err error
+func (l *FileCacheLoader) GetSnapshot(ctx context.Context, keys *fcpb.SnapshotKeys) (*Snapshot, error) {
+	var lastErr error
+	allKeys := append([]*fcpb.SnapshotKey{keys.GetMainKey()}, keys.FallbackKeys...)
+	for i, key := range allKeys {
+		manifest, err := l.getSnapshot(ctx, key)
+		if err != nil {
+			lastErr = err
+			continue
+		}
+		if i == 0 {
+			log.CtxInfof(ctx, "Found preferred snapshot ref=%q in cache", key.GetRef())
+		} else {
+			log.CtxInfof(ctx, "Found fallback snapshot ref=%q in cache", key.GetRef())
```

Comment:
maybe our refs proto should just be a list of refs instead of a preferred + list of refs?

Reply:
I considered that, but kind of like having a separation between the keys that are read+write vs. the keys that are readonly.